File: python/mirage/core/ram/find.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from fnmatch import fnmatch

from mirage.accessor.ram import RAMAccessor
from mirage.types import PathSpec


def _norm(path: str) -> str:
    return "/" + path.strip("/")


async def find(
    accessor: RAMAccessor,
    path: PathSpec,
    name: str | None = None,
    type: str | None = None,
    min_size: int | None = None,
    max_size: int | None = None,
    maxdepth: int | None = None,
    name_exclude: str | None = None,
    or_names: list[str] | None = None,
    iname: str | None = None,
    path_pattern: str | None = None,
    mindepth: int | None = None,
) -> list[str]:
    if isinstance(path, str):
        path = PathSpec(original=path, directory=path)
    if isinstance(path, PathSpec):
        path = path.strip_prefix
    store = accessor.store
    p = _norm(path)
    prefix = p.rstrip("/") + "/"
    base_depth = 0 if p == "/" else p.count("/")
    results: list[str] = []

    candidates: list[tuple[str, str]] = []
    if type != "d":
        for key in store.files:
            candidates.append((key, "f"))
    if type != "f":
        for key in store.dirs:
            if key != "/":
                candidates.append((key, "d"))

    for key, kind in candidates:
        if key != p and not key.startswith(prefix):
            continue
        if key == p and kind == "d":
            continue

        depth = key.count("/") - base_depth

        if maxdepth is not None and depth > maxdepth:
            continue

        if mindepth is not None and depth < mindepth:
            continue

        basename = key.rsplit("/", 1)[-1]
        if name is not None and not fnmatch(basename, name):
            continue

        if iname is not None and not fnmatch(basename.lower(), iname.lower()):
            continue

        if path_pattern is not None and not fnmatch(key, path_pattern):
            continue

        if or_names is not None and not any(
                fnmatch(basename, pat) for pat in or_names):
            continue

        if name_exclude is not None and fnmatch(basename, name_exclude):
            continue

        if kind == "f" and (min_size is not None or max_size is not None):
            size = len(store.files[key])
            if min_size is not None and size < min_size:
                continue
            if max_size is not None and size > max_size:
                continue

        results.append(key)

    return sorted(results)

File: typescript/packages/core/src/core/ram/find.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { RAMAccessor } from '../../accessor/ram.ts'
import type { PathSpec } from '../../types.ts'
import { norm } from './utils.ts'

export interface FindOptions {
  name?: string | null
  type?: 'f' | 'd' | null
  minSize?: number | null
  maxSize?: number | null
  maxDepth?: number | null
  minDepth?: number | null
  nameExclude?: string | null
  orNames?: string[] | null
  iname?: string | null
  pathPattern?: string | null
}

function fnmatch(name: string, pattern: string): boolean {
  let re = '^'
  for (const ch of pattern) {
    if (ch === '*') re += '.*'
    else if (ch === '?') re += '.'
    else if (/[.+^${}()|[\]\\]/.test(ch)) re += '\\' + ch
    else re += ch
  }
  re += '$'
  return new RegExp(re).test(name)
}

export function find(
  accessor: RAMAccessor,
  path: PathSpec,
  options: FindOptions = {},
): Promise<string[]> {
  const p = norm(path.stripPrefix)
  const prefix = p.replace(/\/+$/, '') + '/'
  const baseDepth = p === '/' ? 0 : (p.match(/\//g) ?? []).length
  const results: string[] = []
  const candidates: [string, 'f' | 'd'][] = []
  if (options.type !== 'd') {
    for (const key of accessor.store.files.keys()) candidates.push([key, 'f'])
  }
  if (options.type !== 'f') {
    for (const key of accessor.store.dirs) if (key !== '/') candidates.push([key, 'd'])
  }
  for (const [key, kind] of candidates) {
    if (key !== p && !key.startsWith(prefix)) continue
    if (key === p && kind === 'd') continue
    const depth = (key.match(/\//g) ?? []).length - baseDepth
    if (options.maxDepth !== null && options.maxDepth !== undefined && depth > options.maxDepth)
      continue
    if (options.minDepth !== null && options.minDepth !== undefined && depth < options.minDepth)
      continue
    const basename = key.slice(key.lastIndexOf('/') + 1)
    if (options.name !== null && options.name !== undefined && !fnmatch(basename, options.name))
      continue
    if (
      options.iname !== null &&
      options.iname !== undefined &&
      !fnmatch(basename.toLowerCase(), options.iname.toLowerCase())
    )
      continue
    if (
      options.pathPattern !== null &&
      options.pathPattern !== undefined &&
      !fnmatch(key, options.pathPattern)
    )
      continue
    if (
      options.orNames !== null &&
      options.orNames !== undefined &&
      !options.orNames.some((pat) => fnmatch(basename, pat))
    )
      continue
    if (
      options.nameExclude !== null &&
      options.nameExclude !== undefined &&
      fnmatch(basename, options.nameExclude)
    )
      continue
    if (kind === 'f' && (options.minSize !== null || options.maxSize !== null)) {
      const data = accessor.store.files.get(key)
      if (data === undefined) continue
      const size = data.byteLength
      if (options.minSize !== null && options.minSize !== undefined && size < options.minSize)
        continue
      if (options.maxSize !== null && options.maxSize !== undefined && size > options.maxSize)
        continue
    }
    results.push(key)
  }
  results.sort()
  return Promise.resolve(results)
}
